Notes/BeetleBattleDesign.md

```diff
@@ -0,0 +1,115 @@
+the length of the length of the forwards vector. We can reason likewise for the x-component and cos(θ). Putting this
+together, and realizing that if θ is 90 - `angle`, then it is the other angle in the triangle, and thus we can just use
+`angle` directly:
+
+> `forward_x` = sin(`angle`) \* `forward_speed`
+> `forward_y` = cos(`angle`) \* `forward speed`
+> `sideways_x` = cos(`angle`) \* `sideways_speed`
+> `sideways_y` = sin(`angle`) \* `sideways_speed`
+> `change_x` = `forward_x` + `sideways_x`
+> `change_y` = `forward_y` + `sideways_y`
+
+Thus this is some pseudo-code for handling movement.
+
+### The Projectiles
+
+The projectiles are shot along the facing direction at regular intervals. There is likely a cooldown, the shot is fired,
+and then a bullet particle is shot. If this bullet particle collides with an enemy beetle it does damage. Note, I say "enemy" beetle because I don't think we want friendly fire between the beetles.
+
+These are the weakest attacks, made up by their continuous firing. Stronger attacks exist and are rewarded by beetles
+acquiring traits.
+
+### The Teams
+
+The beetles are divided into two battling teams that start on opposite sides. By default there are three beetles in a
+team, though team traits gained through evolution can add to that number. By default all the beetles in a team are
+equal, though certain team traits may specialize a single beetle with a particularly powerful ability.
+
+The battle ends when all the beetles on either team have been eliminated, making the other team the victor. It could
+technically be possible for the two teams to be eliminated at the same time, but most likely one team over the other
+would "win" based on when processing for each beetle occurs and when victory is awarded.
+
+### The Traits
+
+Each team has a set of traits. These traits apply to the beetles to change their capabilities. These traits may affect
+the entire team of beetles as a whole, give a certain trait to all the beetles in a team, or give a certain trait to
+a single beetle in a team.
+
+Two of these traits are given to each team of beetles at the start of the match, but other traits may be added due
+to progression of the beetles over battles.
+
+### Beetle Specification
+
+Beyond the above which just specifies the basic positioning of the beetle, I'm thinking we may want to create two
+classes, `Beetle` and `Command`. The idea is that the basic stats of a beetle:
+
+* HP
+* Defense
+* Offense
+* Forward Speed
+* Sideways Speed
+
+are contained in the `Beetle` class as attributes, and then attacks and actions the beetle can take are in `Command`,
+following the [Command design pattern](https://gameprogrammingpatterns.com/command.html). So then another attribute of
+the `Beetle` class would be to be able to have a list of commands for it to take queued up in such a way that we can
+have great freedom in designing beetle AI behavior.
```

Comment:
I'm thinking this should be `Action` actually.

Reply:
Also, it may make more sense to model the `Trait` class off of this [Flyweight design pattern](https://gameprogrammingpatterns.com/flyweight.html). I think I may have gotten the two mixed in my head, but both have their use here.

Reply:
Gonna circle back to this when I have a sec to grok this